src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/PostProcessReadsForRSEM.java

```diff
@@ -0,0 +1,257 @@
+ *
+ * This tool reorders such that the alignment pair appears together as in the first example.
+ *
+ * ### Task 2. Removing Reads ###
+ *
+ * If requested, this tool also removes duplicate marked reads and MT reads, which can skew gene expression counts.
+ *
+ *
+ * Caveat: This tool does not remove duplicate reads; it assumes it's been removed upstream e.g.
+ * MarkDuplicates with R
+ *
+ */
+@CommandLineProgramProperties(
+        summary = "",
+        oneLineSummary = "",
+        programGroup = ReadDataManipulationProgramGroup.class // Sato: Change to QC when the other PR is merged.
+)
+public class PostProcessReadsForRSEM extends GATKTool {
+    @Argument(fullName = StandardArgumentDefinitions.OUTPUT_LONG_NAME, shortName = StandardArgumentDefinitions.OUTPUT_SHORT_NAME)
+    public File outSam;
+
+    @Argument(fullName = "keep-MT-reads")
+    public boolean keepMTReads = false;
+
+    PeekableIterator<GATKRead> read1Iterator;
+    SAMFileGATKReadWriter writer;
+
+    ReadPair currentReadPair;
+
+    // Use CountingFilter or another existing framework for counting and printing filtered reads
+    int totalOutputPair = 0;
+    int notBothMapped = 0;
+    int chimera = 0;
+    int unsupportedCigar = 0;
+    int duplicates = 0;
+    int mitochondria = 0;
+
+    @Override
+    public void onTraversalStart(){
+        read1Iterator = new PeekableIterator<>(directlyAccessEngineReadsDataSource().iterator());
+
+        if (directlyAccessEngineReadsDataSource().getHeader().getSortOrder() != SAMFileHeader.SortOrder.queryname){
+            throw new UserException("Input must be query-name sorted.");
+        }
+
+        writer = createSAMWriter(new GATKPath(outSam.getAbsolutePath()), true);
+
+        if (!read1Iterator.hasNext()){
+            throw new UserException("Input has no reads");
+        }
+
+        currentReadPair = new ReadPair(read1Iterator.next());
+        totalOutputPair++;
+    }
+
+    @Override
+    public void traverse() {
+        while (read1Iterator.hasNext()){
+            final GATKRead read = read1Iterator.next();
+            if (!currentReadPair.getQueryName().equals(read.getName())){
+                // We have gathered all the reads with the same query name (primary, secondary, and supplementary alignments)
+                // Write the reads to output file, reordering the reads as needed.
+                writeReads(currentReadPair);
+                currentReadPair = new ReadPair(read);
+            } else {
+                currentReadPair.add(read);
+            }
+            progressMeter.update(read);
+        }
+    }
+
+    public boolean passesRSEMFilter(final GATKRead read1, final GATKRead read2) {
+        // If either of the pair is unmapped, throw out.
+        // With the STAR argument --quantTranscriptomeBan IndelSoftclipSingleend this should not occur,
+        // but we check just to be thorough.
+        if (read1.getContig() == null || read2.getContig() == null){
```

Comment:
If either read1 or read2 is missing this will crash with a NPE.  That's probably something that should be handled since it tends to happen.  A crash is a fine answer but detect it and write a sane error message.

Since it seems like you do require both a first and second in pair, one way to simplify some things would be to push these checks into ReadFilters and then at this point you'll just check if there is a full set of reads.   It would change the counting of the elimination reason though from being per fragment group to being by read which might not be what you want.

Reply:
This doesn't seem to happen. But I'm going to invoke the @droazen rule and do the pragmatic thing, which is that if either read1 or read2 is null, we return false (so these reads will not be output).